### PR TITLE
Use line numbers relative to the function in mir-opt tests

### PR DESCRIPTION
**Pull request by Noratrieb**
As shown in #99770, the line numbers can be a big source of needless and confusing diffs. This PR adds a new flag `-Zmir-pretty-relative-line-numbers` to make them relative to the function declaration, which avoids most needless diffs from attribute changes.

@JakobDegen told me that there has been a zulip conversation about disabling line numbers with mixed opinions, so I'd like to get some feedback here, for this hopefully better solution.

r? rust-lang/wg-mir-opt

### CODE REVIEW

**Comment by Nilstrieb** [comment]
r? rust-lang/mir-opt

**Comment by klensy** [comment]
Maybe reuse already existing style like `LL:COL` instead of `$LINE:$COL`?:
https://github.com/rust-lang/rust/blob/c11207ec89b856164bba03b8ecfe07b0b234ed21/src/test/mir-opt/const_prop/control_flow_simplification.hello.ConstProp.diff#L18

**Comment by oli-obk** [comment]
Maybe we could make this info relative to the function start? That should avoid most of the noise, right?

**Comment by Nilstrieb** [comment]
This should avoid most noise, yes. It would probably be a bit tricky to implement with the current code structure, but I could do it. What do you think would be the best way to solve this? Rename -Zmir-pretty-line-numbers`  to something like `-Zmir-pretty-relative-line-numbers` and adding a new method on `source_map` for relative line numbers? The span of the function should be available in mir pretty printing I think.

**Comment by oli-obk** [comment]
> Rename -Zmir-pretty-line-numbers` to something like`-Zmir-pretty-relative-line-numbers`and adding a new method on`source_map` for relative line numbers?

that sounds reasonable to me

File: compiler/rustc_span/src/source_map.rs [hunk 1]
```diff
@@ -463,6 +463,29 @@ impl SourceMap {
         self.span_to_string(sp, FileNameDisplayPreference::Remapped)
     }
 
+    /// Format the span location suitable for pretty printing anotations with relative line numbers
+    pub fn span_to_relative_line_string(&self, sp: Span, relative_to: Span) -> String {
+        if self.files.borrow().source_files.is_empty() || sp.is_dummy() || relative_to.is_dummy() {
+            return "no-location".to_string();
+        }
+
```
Comment (oli-obk) [inline compiler/rustc_span/src/source_map.rs:471]: I think we should `assert!(relative_to.contains(sp))`

File: compiler/rustc_span/src/source_map.rs [hunk 2]
```diff
@@ -463,6 +463,29 @@ impl SourceMap {
...
+        let hi_line = hi.line.saturating_sub(offset.line);
+
+        format!(
+            "{}:{}:{}: {}:{}",
+            lo.file.name.display(FileNameDisplayPreference::Remapped),
+            lo_line,
+            lo.col.to_usize() + 1,
+            hi_line,
+            hi.col.to_usize() + 1,
+        )
```
Comment (oli-obk) [inline compiler/rustc_span/src/source_map.rs:486]: In order for it to be clearer that these are offsets, add a `+` before every line number.

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_span/src/source_map.rs (hunk 1 above)
Comment (Nilstrieb) [reply compiler/rustc_span/src/source_map.rs:471]: In the way it's currently used the assert would make sense, but I think usage where the two spans are separate but one is before the other would also be correct. The assert would panic though.

Comment (oli-obk) [reply compiler/rustc_span/src/source_map.rs:471]: yea that's true, but I think we'll also want to know if anyone starts doing that. I don't know how well that would be supported right now, so I'd rather not silently support it and get suprised in the future.

**Comment by Nilstrieb** [comment]
Oops, turns out there was a bug that the assert uncovered. Are you able to stop the bors run since it will fail anyways? @oli-obk 

I found the cause and I'll push a fix soon.

**Comment by Nilstrieb** [comment]
Hmm, there's a problem with associated consts (and maybe more, haven't checked) where the bb stmts of the const have a span that's outside their `body.span`. That doesn't seem ideal, but I don't think we should block this on that. I think I'll just remove the assert again instead of trying to make the spans for MIR perfectly correct everywhere here.

There was also another bug if the two spans are from different files - I fixed that.

**Comment by oli-obk** [comment]
@bors r+ p=1

yea let's get this change done first and then fine tune the (significantly fewer) cases that need some adjusting

**Comment by Nilstrieb** [comment]
Coverage, yay. I first need to get that working, which I wasn't able to myself, so I asked on zulip for help.

**Comment by Nilstrieb** [comment]
I figured it out. Maybe it works now.

**Comment by Nilstrieb** [comment]
I think you're a bit late my friend, but thank you for your log analysis anyways :)

**Pull request merged**